config/v1/types_infrastructure.go

```diff
@@ -77,8 +77,34 @@ type InfrastructureStatus struct {
 	// like kubelets, to contact the Kubernetes API server using the
 	// infrastructure provider rather than Kubernetes networking.
 	APIServerInternalURL string `json:"apiServerInternalURI"`
+
+	// ControlPlaneTopology expresses the expectations for operands that normally run on control nodes.
+	// The default is 'HighlyAvailable', which represents the behavior operators have in a "normal" cluster.
+	// The 'SingleReplica' mode will be used in single-node deployments (developer and production) for example,
+	// and the operators should not configure the operand for highly-available operation
+	// +kubebuilder:default=HighlyAvailable
+	ControlPlaneTopology TopologyMode `json:"controlPlaneTopology"`
+
+	// InfrastructureTopology expresses the expectations for operands that normally run on infrastructure nodes.
```

Comment:
"infrastructure nodes" are an OSD thing, not an OCP thing, and this is an OCP config structure.  I realize that the enhancement discussed this addition, and am sorry I missed this in review there.  Thoughts about revisiting in the enhancement, or discussing here, or...?  I guess this API change will not land before 4.7 splits off master anyway, so we have some time...

Reply:
Where else would we put the new field?

This was added at @derekwaynecarr's suggestion for the external control plane work, which makes me think the idea of infrastructure nodes is coming to OCP?

Reply:
@derekwaynecarr proposing a control-plane/infra split [here][1], with an eye to [future `External` control plane settings][2] or some such.  That seems cross-cutty to me.  As long as the only two values are `HighlyAvailable` and `SingleReplica`, I can't think of a situation where folks would set one value for `ControlPlaneTopology` and another for `InfrastructureTopology`.  If we eventually grow `External`, the distinction would be:

> `ControlPlaneTopology` stops being a way to tell control-plane components how available to be, and becomes a way to tell the remaining components that the control plane components are not local.

I'd rather have addressed that with a cluster-profile knob, since we're using the cluster-profile knob to actually remove the control-plane components.  As it stands, which way would, say, the registry operator go?  It [tolerates the control-plane nodes][3], but also [stays local, even in the IBM/ROKS case where the bulk of the control-plane moves to external hosting][4].  Maybe all the components that fall into that category won't care about this setting anyway (anything managed by the CVO, which includes the registry-operator manifest, will not care about these settings).  But it seems like "second-level operators should consume `InfrastructureTopology` when they are running, and look to `ControlPlaneTopology` in the future if it grows `External`?  That makes the names a bit wonky, but seems like the logic folks were going for?

[1]: https://github.com/openshift/enhancements/pull/555#discussion_r553574042
[2]: https://github.com/openshift/enhancements/pull/555#discussion_r553574576
[3]: https://github.com/openshift/cluster-image-registry-operator/blob/67118811beaa9317b1f705233417bfadb7b34462/manifests/07-operator.yaml#L28-L30
[4]: https://github.com/openshift/cluster-image-registry-operator/blob/67118811beaa9317b1f705233417bfadb7b34462/manifests/07-operator.yaml#L9

Reply:
IIUC the goal for the registry operator is to leverage the cluster profiles to support the IBM/ROKS use case:
https://github.com/openshift/cluster-image-registry-operator/pull/643
We (monitoring team) have a similar WIP [PR](https://github.com/openshift/cluster-monitoring-operator/pull/986) from @csrwng to move the cluster monitoring operator from the control plane nodes to the worker nodes.

Reply:
@wking @simonpasquier I think we can draw a line between what cluster profiles do and what these fields are meant to accomplish. Cluster profiles affect the manifests that get laid down for operators by the CVO. In the cases that @simonpasquier pointed out, there is no opportunity to change the node selectors via operator logic because they will be laid down directly by the CVO, so if we need alternate node selectors we have to have the CVO handle it.

Cluster profiles *could* also be used to drive logic in the operator such as adding an environment variable or some other change to an operator manifest to indicate that it should work in HA/non-HA mode. But IMHO it would be less than ideal because then each operator gets to decide how this is communicated to them and they may not agree on what each mode means. My understanding is that these fields are a consistent way to communicate to operators the desired availability mode.

For possible combinations see https://github.com/openshift/enhancements/pull/555#discussion_r553994319

Reply:
@wking @simonpasquier The reason for not exposing the profile name directly through the API is that it means every time we create a new profile, we automatically have to update every operator that looks at the field to be aware of how to behave under that profile. See  [this section of the enhancement doc](https://github.com/openshift/enhancements/blob/master/enhancements/cluster-high-availability-mode-api.md#expose-the-cluster-profile-name-directly). By decomposing that "high level" concept into smaller parts, we only have to update operators that care about the smaller thing. Right now, everything that cares about the "profile" will care about one or the other of these topology fields, but not both. Eventually, as we add more similar fields we anticipate only a subset of the operators caring about those as well, so each change only touches the components that need to be updated.

@csrwng We drew a somewhat firmer line in the discussions we had about whether to use profiles for single node at all. Profiles cause components to be included in or excluded from the cluster. APIs are used to configure components differently within the cluster. It seems to me if we need to change a Deployment based on an API, that Deployment needs to be managed by an operator, even if the component inside the Deployment is itself an operator. So, we might end up introducing another layer between the CVO and some operator(s) that need to be configured with different node selectors based on the cluster topology API(s).

Reply:
I buy the general argument for config knobs.  I just think that `InfrastructureTopology` is an ambiguous config knob.  If everyone else feels like it's pretty clear which components are controlled by `ControlPlaneTopology` and which are controlled by `InfrastructureTopology`, then feel free to ignore me and keep this rolling along.

Reply:
@derekwaynecarr suggested the split. I'm not 100% clear on what goes on each side of the line myself. We should probably go back and document that better in the enhancement (and in comments here).

Reply:
@wking this is not to be read as literal as role=infra nodes.  it should be understood as components that provide supporting infrastructure services for applications on the cluster but are not run on control plane hosts by default.  for example, it would include components like the registry, router, and monitoring.  today those components can be selected to run on node with role=infra but they default to role=worker.  what this field is intending to capture is if those supporting services should run highly available or not.  it could be read by additional add-ons as a hint in the future for kubevirt, serverless, mesh, pipelines, or any other supporting infrastructure service to understand what availability posture their supporting solution should provide by default.

a very simple use case is a cluster whose control plane is externalized, but whose infrastructure services do not require HA.  for example, an IBM ROKS/Satellite cluster with a single worker.  this would require a router to not error when its not running HA, which this flag would signal as the desired state.

so maybe update text as follows:

```
InfrastructureTopology expresses the expectations for operands that provide supporting infrastructure services for workloads on the cluster, but do not run on control hosts by default.  If an operator API provides a node selector for placement of its operands, the operator should respect this API for expressing its availability absent any local override.
```

Reply:
an alternative scenario is regular standalone highly-available topology (3 control plane hosts) with a single worker.  in that model, i may desire HA control plane topology, and SingleReplica Infrastructure Topology.

Reply:
I've proposed an update to the original API enhancement to clarify this. See https://github.com/openshift/enhancements/pull/605

Reply:
I updated the comment in the code accordingly.